### ISSUE TITLE
Do not pipe if hint is cancelled?

### ISSUE THREAD

**Issue by hejops** [tridactyl/tridactyl#2920]
*   Brief description of the problem:

I'd like to know if a pipe can be "terminated" if I decide to cancel a hint.

*   Steps to reproduce:

1. `composite hint -y | js -p tri.excmds.shellescape(JS_ARG).then(url => tri.excmds.exclaim_quiet('chromium ' + url))` (or any appropriate executable)
2. Cancel the hint (Esc)
3. Chromium is still executed

*   Unix-like only: result of running `printf '%c\0\0\0{"cmd": "run", "command": "echo $PATH"}' 39 | ~/.local/share/tridactyl/native_main.py` in a terminal:

```ø{"cmd": "run", "content": "/home/joseph/.local/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/sbin:/usr/lib/jvm/default/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl:/home/joseph/scripts:/home/joseph/.gem/ruby/2.7.0/bin\n", "code": 0}```

**Comment by bovine3dom** [comment]
No, we don't have any concept of success or failure, so once a pipe has started it can't be stopped. However, you can probably do what you want with `hint -W [ex command]` - e.g.

```
command openinchrome js -p tri.excmds.shellescape(JS_ARG).then(url => tri.excmds.exclaim_quiet('chromium ' + url))
hint -W openinchrome
```

**Comment by bovine3dom** [comment]
Another solution is to check that `url` isn't empty in your JS snippet

**Comment by hejops** [comment]
`hint -W openinchrome` works quite well, thank you.

Can the -W flag be combined with -Jc? I tried something like `hint -Jc <elements> -W openinchrome` but the hint fails.

**Comment by bovine3dom** [comment]
No, it can't. Hint mode is very messy really. You could add it or you could switch to callback hint mode - `:hint -JFc [selector with no spaces] elem => [some js]`. https://github.com/tridactyl/tridactyl/blob/eca1e1614b8d0e51dcb7369a3b4026fed72ad209/src/excmds.ts#L3998

To add it to W, take the withSelectors part from here https://github.com/tridactyl/tridactyl/blob/eca1e1614b8d0e51dcb7369a3b4026fed72ad209/src/excmds.ts#L4322 and replace https://github.com/tridactyl/tridactyl/blob/eca1e1614b8d0e51dcb7369a3b4026fed72ad209/src/excmds.ts#L4186 with it.

**Comment by hejops** [comment]
I'll see what I can do, thanks!
